packages/backend-data/src/factory.ts

```diff
@@ -171,15 +177,57 @@ class DataGenerator implements ConstructContainerEntryGenerator {
       authorizationModes,
       outputStorageStrategy: this.outputStorageStrategy,
       functionNameMap,
-      translationBehavior: { sandboxModeEnabled },
+      translationBehavior: {
+        sandboxModeEnabled,
+        /**
+         * The destructive updates should be always allowed in backend definition and not to be controlled on the IaC
+         * The CI/CD check should take the responsibility to validate if any tables are being replaced and determine whether to execute the changeset
+         */
+        allowDestructiveGraphqlSchemaUpdates: true,
+      },
     });
+    Aspects.of(amplifyApi).add(new ReplaceTableUponGsiUpdateChecker());
 
     convertJsResolverDefinition(scope, amplifyApi, jsFunctions);
 
     return amplifyApi;
   };
 }
 
+const REPLACE_TABLE_UPON_GSI_UPDATE_ATTRIBUTE_NAME: keyof TranslationBehavior =
+  'replaceTableUponGsiUpdate';
+
+enum BackendDataEnvironmentVariables {
+  // Environment variables allowed in sandbox mode only
+  REPLACE_TABLE_UPON_GSI_UPDATE = 'REPLACE_TABLE_UPON_GSI_UPDATE',
+}
+/**
+ * Aspect class to modify the amplify managed DynamoDB table
+ * to allow table replacement upon GSI update when the environment variable is set to true
+ * This is only allowed to be modified in sandbox mode
+ */
+class ReplaceTableUponGsiUpdateChecker implements IAspect {
+  public visit(scope: IConstruct): void {
+    const replaceTableUponGsiUpdate: boolean =
+      process.env[
+        BackendDataEnvironmentVariables.REPLACE_TABLE_UPON_GSI_UPDATE
```

Comment:
Another option: the data construct could read from a CDK context value to determine this behavior. Then here we only need to set the CDK context value rather than have this aspect override.

Even better would be to expose this as a normal prop on the data construct that we can set rather than have this "backdoor"

Reply:
I think it is better not to leave this param in the data construct as it is a param we only want to take effect in the sandbox mode and it should never happen in other type of deployment. The main purpose for this flag is to iterate fast deployment with the sacrifice of the table data, which is not suitable for branch deployment.

For the idea of using CDK context value, it was adopted in a [previous PR](https://github.com/aws-amplify/amplify-category-api/pull/2078) but it would pass the reading context value logic into the data construct. It was not a desired option at that time.

Reply:
Confirmed from product: always allow GSI update replacing table in sandbox